Firestore/core/test/unit/local/remote_document_cache_test.cc

```diff
@@ -230,34 +232,85 @@ TEST_P(RemoteDocumentCacheTest, DocumentsMatchingUsesReadTimeNotUpdateTime) {
+
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    MutableDocumentMap documents =
+        cache_->GetAll(DocumentKeySet{Key("coll/doc")});
+    document = documents.find(Key("coll/doc"))->second;
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    documents = cache_->GetMatching(Query("coll"), SnapshotVersion::None());
+    document = documents.find(Key("coll/doc"))->second;
+    VerifyValue(document, Map("value", "old"));
+    document.ConvertToFoundDocument(Version(kVersion),
+                                    ObjectValue{Map("value", "new")});
+
+    document = cache_->Get(Key("coll/doc"));
+    VerifyValue(document, Map("value", "old"));
+  });
+}
 // MARK: - Helpers
 
-Document RemoteDocumentCacheTest::SetTestDocument(const absl::string_view path,
-                                                  int update_time,
-                                                  int read_time) {
-  Document doc = Doc(path, update_time, kDocData);
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path,
+    google_firestore_v1_Value data,
+    int update_time,
+    int read_time) {
+  MutableDocument doc = Doc(path, update_time, data);
   cache_->Add(doc, Version(read_time));
   return doc;
 }
 
-Document RemoteDocumentCacheTest::SetTestDocument(
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path, int update_time, int read_time) {
+  return SetTestDocument(path, DeepClone(kDocData), update_time, read_time);
+}
+
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path, google_firestore_v1_Value data) {
+  return SetTestDocument(path, data, kVersion, kVersion);
+}
+
+MutableDocument RemoteDocumentCacheTest::SetTestDocument(
     const absl::string_view path) {
-  return SetTestDocument(path, kVersion, kVersion);
+  return SetTestDocument(path, DeepClone(kDocData), kVersion, kVersion);
+}
+
+void RemoteDocumentCacheTest::VerifyValue(MutableDocument actual_doc,
+                                          google_firestore_v1_Value data) {
+  MutableDocument expected_doc =
+      Doc(actual_doc.key().ToString(), kVersion, data);
+  ASSERT_EQ(expected_doc, actual_doc);
 }
 
 void RemoteDocumentCacheTest::SetAndReadTestDocument(
     const absl::string_view path) {
   persistence_->Run("SetAndReadTestDocument", [&] {
-    Document written = SetTestDocument(path);
-    absl::optional<MaybeDocument> read = cache_->Get(testutil::Key(path));
+    MutableDocument written = SetTestDocument(path);
+    absl::optional<MutableDocument> read = cache_->Get(Key(path));
     ASSERT_EQ(*read, written);
```

Comment:
Same comments as `VerifyValue`.

Reply:
Used EXPECT_EQ. Leaving the Scoped Trace portion as an exercise to the reader as this is existing code.